### PR TITLE
[SDP-1397] Circle Payouts pt2: error handling & retries

### PR DESCRIPTION
**Pull request by marcelosalloum**
### What

Errors in the POST /recipient endpoint are not being properly handled. What;s missing?
- When POST /recipient fails with `ErrCircleRecipientCreationFailedTooManyTimes`, we update the payments table with the failed status explaining the reason.
- When a payment that failed due to recipient creation issues is retried, the recipient is reset with `status=NULL` and `sync_attempts=0`.
- The method `ensureRecipientIsReady` was split into different sub-methods for better modularity and clarity, as suggested in https://github.com/stellar/stellar-disbursement-platform-backend/pull/486#discussion_r1878843074.

Screenshot 2024-12-12 at 4 39 08 PM

- [x] Tested with scheduled jobs
- [x] Tested with Kafka

### Why

To wrap up https://stellarorg.atlassian.net/browse/SDP-1397

#### PR Structure

* [x] This PR has a reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR title and description are clear enough for anyone to review it.
* [x] This PR does not mix refactoring changes with feature changes (split into two PRs otherwise).

#### Thoroughness

* [x] This PR adds tests for the new functionality or fixes.
* [x] This PR contains the link to the Jira ticket it addresses.

#### Configs and Secrets

* [x] No new CONFIG variables are required -OR- the new required ones were added to the helmchart's [`values.yaml`] file.
* [x] No new CONFIG variables are required -OR- the new required ones were added to the deployments ([`pr-preview`], [`dev`], [`demo`], `prd`).
* [x] No new SECRETS variables are required -OR- the new required ones were mentioned in the helmchart's [`values.yaml`] file.
* [x] No new SECRETS variables are required -OR- the new required ones were added to the deployments ([`pr-preview secrets`], [`dev secrets`], [`demo secrets`], `prd secrets`).

#### Release

* [x] This is not a breaking change.
* [x] **This is ready for production.**. If your PR is not ready for production, please consider opening additional complementary PRs using this one as the base. Only merge this into `develop` or `main` after it's ready for production!

#### Deployment

* [x] Does the deployment work after merging?

[`values.yaml`]: ../helmchart/sdp/values.yaml
[`pr-preview`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/stellar-disbursement-platform/backend-helm-values
[`dev`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/backend-helm-values
[`demo`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-backend-helm-values
[`pr-preview secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/common-previews/externalsecrets-common-previews.yaml#L241-L346
[`dev secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/stellar-disbursement-platform-externalsecrets.yaml
[`demo secrets`]: https://github.com/stellar/kube/blob/d3e4f5dd8aa4c13b45a31a5a937f3e98841171a7/kube001-dev/namespaces/stellar-disbursement-platform/demo/demo-sdp-externalsecrets.yaml

### CODE REVIEW

File: internal/services/paymentdispatchers/circle_payment_dispatcher.go [hunk 1]
```diff
@@ -212,6 +228,15 @@ func (c *CirclePaymentDispatcher) ensureRecipientIsReady(ctx context.Context, re
 		},
 	})
 	if err != nil {
+		// Bump the sync_attempt count if the recipient creation failed
+		_, updateErr := c.sdpModels.CircleRecipient.Update(ctx, dataRecipient.ReceiverWalletID, data.CircleRecipientUpdate{
+			SyncAttempts:      dataRecipient.SyncAttempts + 1,
+			LastSyncAttemptAt: time.Now(),
```
Comment (marwen-abid) [inline internal/services/paymentdispatchers/circle_payment_dispatcher.go:234]: Should we save the error in the ResponseBody ?

File: internal/services/paymentdispatchers/circle_payment_dispatcher.go [hunk 2]
```diff
@@ -224,36 +249,65 @@ func (c *CirclePaymentDispatcher) ensureRecipientIsReady(ctx context.Context, re
...
+		initialDelay = 100 * time.Millisecond
+	}
+
 	var recipient *data.CircleRecipient
 	err := retry.Do(
 		func() error {
 			var err error
 			recipient, err = c.ensureRecipientIsReady(ctx, receiverWallet)
 			if err != nil {
 				if errors.Is(err, ErrCircleRecipientCreationFailedTooManyTimes) {
```
Comment (marwen-abid) [inline internal/services/paymentdispatchers/circle_payment_dispatcher.go:308]: ❓I wonder if we will ever really reach this condition if we already have the `retry.Attempts(maxCircleRecipientCreationAttempts)`. 

If I understand this correctly : this is how a creation failure would go : 
1. Submit Circle Recipient 
2. Sync -> Recipient is Pending (retry 1)
3. Sync -> Recipient Error -> Try again (retry 2, SyncAttempts 1)
4. Sync -> Recipient Error -> Try again (retry 3, SyncAttempts 2)
5. Sync -> Recipient Error -> Try again (retry 4, SyncAttempts 3)
6. Sync -> Recipient Error -> Try again (retry 5, SyncAttempts 4) -> Abort. 

So maybe we should create two constants : 
* maxCircleRecipientCreationAttempts = 5
* maxCircleRecipientSyncAttempts > maxCircleRecipientCreationAttempts

**Review by marwen-abid — APPROVED** [review]
LGTM! This is great!

File: internal/services/paymentdispatchers/circle_payment_dispatcher.go [hunk 3]
```diff
@@ -63,46 +63,57 @@ var _ PaymentDispatcherInterface = (*CirclePaymentDispatcher)(nil)
 
 func (c *CirclePaymentDispatcher) sendPaymentsToCircle(ctx context.Context, sdpDBTx db.DBTransaction, circleWalletID string, paymentsToSubmit []*data.Payment) error {
 	for _, payment := range paymentsToSubmit {
-		// 0. Ensure the recipient is ready
-		recipient, err := c.ensureRecipientIsReadyWithRetry(ctx, *payment.ReceiverWallet)
+		// 1. Ensure the recipient is ready
+		// TODO: When clicking retries, we should reset the recipient count?
```
[the remark this answers is not in the document]
Comment (marcelosalloum) [reply internal/services/paymentdispatchers/circle_payment_dispatcher.go]: Thanks for pointing that out! Addressed in https://github.com/stellar/stellar-disbursement-platform-backend/pull/491/commits/67f1b4d208313a84fd60a7d8ced6da4ddab2f1c8.

File: internal/services/paymentdispatchers/circle_payment_dispatcher.go [hunk 4]
```diff
@@ -63,46 +63,57 @@ var _ PaymentDispatcherInterface = (*CirclePaymentDispatcher)(nil)
...
+		recipient, err := c.ensureRecipientIsReadyWithRetry(ctx, *payment.ReceiverWallet, 100*time.Millisecond)
```
[the remark this answers is not in the document]
Comment (marcelosalloum) [reply internal/services/paymentdispatchers/circle_payment_dispatcher.go]: Addressed in https://github.com/stellar/stellar-disbursement-platform-backend/pull/491/commits/cbc5c18214029683966ac70e8833bcc21cbf4696

File: internal/serve/httphandler/payments_handler.go [hunk 5]
```diff
@@ -178,6 +178,11 @@ func (p PaymentsHandler) RetryPayments(rw http.ResponseWriter, req *http.Request
 				return nil, fmt.Errorf("retrying failed payments: %w", err)
 			}
 
+			_, err = p.Models.CircleRecipient.ResetRecipientsForRetryIfNeeded(ctx, dbTx, reqBody.PaymentIDs...)
+			if err != nil {
+				return nil, fmt.Errorf("resetting circle recipients for retry if needed: %w", err)
+			}
```
[the remark this answers is not in the document]
Comment (marcelosalloum) [reply internal/serve/httphandler/payments_handler.go:190]: Updated in https://github.com/stellar/stellar-disbursement-platform-backend/pull/491/commits/83469c5c1e2a3b7b9ffaa16a81b35c4a10d62c68.

File: internal/services/paymentdispatchers/circle_payment_dispatcher.go (hunk 2 above)
Comment (marcelosalloum) [reply internal/services/paymentdispatchers/circle_payment_dispatcher.go:314]: We could rely solely on the `ErrCircleRecipientCreationFailedTooManyTimes` being returned if we set the retry limit to `0` (considered unlimited retries by the lib).

The main difference would be:
- `retry.Attempts(maxCircleRecipientCreationAttempts)`: displays the aggregated errors from all failed (re)tries.
- `retry.Attempts(0)`: only displays the latest (re)tried error.

Comment (marwen-abid) [reply internal/services/paymentdispatchers/circle_payment_dispatcher.go:314]: Thank you for the explanation! Makes sense

**Pull request merged**